src/io/platform/f3xx/f302x8/SPIf302x8.cpp

```diff
@@ -0,0 +1,359 @@
+    } else if (baudRate >= SPI_MAX_BAUD / 2) {
+        halSPI.Init.BaudRatePrescaler = SPI_BAUDRATEPRESCALER_4;
+    } else if (baudRate >= SPI_MAX_BAUD / 4) {
+        halSPI.Init.BaudRatePrescaler = SPI_BAUDRATEPRESCALER_8;
+    } else if (baudRate >= SPI_MAX_BAUD / 8) {
+        halSPI.Init.BaudRatePrescaler = SPI_BAUDRATEPRESCALER_16;
+    } else if (baudRate >= SPI_MAX_BAUD / 16) {
+        halSPI.Init.BaudRatePrescaler = SPI_BAUDRATEPRESCALER_32;
+    } else if (baudRate >= SPI_MAX_BAUD / 32) {
+        halSPI.Init.BaudRatePrescaler = SPI_BAUDRATEPRESCALER_64;
+    } else if (baudRate >= SPI_MAX_BAUD / 64) {
+        halSPI.Init.BaudRatePrescaler = SPI_BAUDRATEPRESCALER_128;
+    } else {
+        halSPI.Init.BaudRatePrescaler = SPI_BAUDRATEPRESCALER_256;
+    }
+
+    // configure the bit order of the data; MSB or LSB
+    if (order) {
+        halSPI.Init.FirstBit = SPI_FIRSTBIT_MSB;
+    } else {
+        halSPI.Init.FirstBit = SPI_FIRSTBIT_LSB;
+    }
+
+    HAL_SPI_Init(&halSPI);
+}
+
+/**
+     * toggles a GPIO pin's state
+     * @param pin the pin to toggle the state of
+     */
+void SPIf302x8::togglePin(GPIO* pin) {
+    switch (pin->readPin()) {
+    case GPIO::State::HIGH:
+        pin->writePin(GPIO::State::LOW);
+        break;
+    case GPIO::State::LOW:
+        pin->writePin(GPIO::State::HIGH);
+        break;
+    }
+}
+
+/**
+     * toggle the state of the chip select pin of a device at the start of a transmission.
+     * @param device device the device index in the CSPins
+     * @return true if valid device, false if device not in CSPins
+     */
+bool SPIf302x8::startTransmission(uint8_t device) {
+    if (device < CSPinsLength) {
+        togglePin(&CSPins[device]);
+        return true;
+    }
+    return false;
+}
+
+/**
+     * toggle the state of the chip select pin of a device back at the end of a transmission.
+     * @param device the device index in the CSPins
+     * @return true if valid device, false if device not in CSPins
+     */
+bool SPIf302x8::endTransmission(uint8_t device) {
+    if (device < CSPinsLength) {
+        togglePin(&CSPins[device]);
+        return true;
+    }
+    return false;
+}
+
+/**
+     * Writes a single byte out to the SPI device
+     * @param byte the byte to write
+     */
+void SPIf302x8::write(uint8_t byte) {
+    HAL_SPI_Transmit(&halSPI, &byte, 1, DEFAULT_SPI_TIMEOUT);
+}
+
+/**
+     * reads a single byte from a SPI device
+     * @return the byte read
+     */
+uint8_t SPIf302x8::read() {
+    uint8_t data;
+    HAL_SPI_Receive(&halSPI, &data, 1, DEFAULT_SPI_TIMEOUT);
+    return data;
+}
+
+/**
+     * writes an array of bytes to the SPI device
+     * @param device the device to write to in CSPins
+     * @param bytes an array of bytes of length n to write to SPI device
+     * @param length the length of the array
+     */
+void SPIf302x8::write(uint8_t* bytes, uint8_t length) {
+    //if (startTransmission(device)) {
```

Comment:
If you don't need the code anymore, remember to delete commented out code (or if it's important to have the commented code for later implementation, add comments explaining it.

Reply:
Just missed that in my cleaning up!